homeassistant/components/weather/intent.py

```diff
@@ -0,0 +1,87 @@
+"""Intents for the weather integration."""
+from __future__ import annotations
+
+import voluptuous as vol
+
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers import intent
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.entity_component import EntityComponent
+
+from . import DOMAIN, WeatherEntity
+
+INTENT_GET_WEATHER = "HassGetWeather"
+
+
+async def async_setup_intents(hass: HomeAssistant) -> None:
+    """Set up the weather intents."""
+    intent.async_register(hass, GetWeatherIntent())
+
+
+class GetWeatherIntent(intent.IntentHandler):
+    """Handle GetWeather intents."""
+
+    intent_type = INTENT_GET_WEATHER
+    slot_schema = {vol.Optional("name"): cv.string}
+
+    async def async_handle(self, intent_obj: intent.Intent) -> intent.IntentResponse:
+        """Handle the intent."""
+        hass = intent_obj.hass
+        slots = self.async_validate_slots(intent_obj.slots)
+
+        weather: WeatherEntity | None = None
+        component: EntityComponent[WeatherEntity] = hass.data[DOMAIN]
+        entities = list(component.entities)
+
+        if "name" in slots:
+            # Named weather entity
+            weather_name = slots["name"]["value"]
+
+            # Find matching weather entity
+            matching_states = intent.async_match_states(
+                hass, name=weather_name, domains=[DOMAIN]
+            )
+            for weather_state in matching_states:
+                for maybe_weather in entities:
```

Comment:
Don't iterate through `entities`, instead use `component.get_entity(weather_state.entity_id)`

Reply:
Like this https://github.com/home-assistant/core/pull/102701/files#r1371010203